src/auth/__tests__/AuthProvider.unit.test.tsx

```diff
@@ -0,0 +1,428 @@
+    });
+
+    it("does not change AuthProvider internal state if result of check is the same as current state", () => {
+      // (1) Check anonymous users
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: null } });
+      });
+
+      auth = result.current;
+      auth.check(true);
+
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: null } });
+      });
+
+      expect(auth).toBe(result.current);
+
+      // (2) Check logged-in users
+      auth.check(true);
+
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: testUser } });
+      });
+
+      auth = result.current;
+      auth.check(true);
+
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: testUser } });
+      });
+
+      expect(auth).toBe(result.current);
+    });
+
+    it("does not change AuthProvider internal state until promise is resolved", () => {
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: null } });
+      });
+
+      auth = result.current;
+      auth.check(true);
+
+      expect(auth).toBe(result.current);
+
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: testUser } });
+      });
+
+      expect(auth).not.toBe(result.current);
+    });
+
+    it("updates the user if they logged in between checks", () => {
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: null } });
+      });
+
+      auth = result.current;
+      auth.check(true);
+
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: testUser } });
+      });
+
+      expect(auth).not.toBe(result.current);
+
+      auth = result.current;
+      expect(auth.user).not.toBeNull();
+    });
+
+    it("updates the user if they logged out between checks", () => {
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: testUser } });
+      });
+
+      auth = result.current;
+      auth.check(true);
+
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: null } });
+      });
+
+      expect(auth).not.toBe(result.current);
+
+      auth = result.current;
+      expect(auth.user).toBeNull();
+    });
+
+    it("updates the user if they changed between checks", () => {
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: testUser } });
+      });
+
+      auth = result.current;
+      auth.check(true);
+
+      act(() => {
+        mockAxios.mockResponse({ data: { success: true, data: testAdmin } });
+      });
+
+      expect(auth).not.toBe(result.current);
+
+      auth = result.current;
+      expect(auth.user).not.toEqual(testUser);
+    });
+
+    it("handles unsuccessful requests by throwing an error", async () => {
+      let checkPromise: Promise<any>;
+
+      checkPromise = auth.check(true)!; // if the force flag is true then check must return a promise
```

Comment:
Not quite sure what's happening here: according to the types, `check` always returns a `Promise<void>`.

Reply:
That was before I changed the types so that that would be true. I've since fixed the test.